### PR TITLE
code updated for implementation of AssetBundle

### PR DESCRIPTION
**Pull request by AyushVerma2**
code updated for implementation of AssetBundle

### CODE REVIEW

File: src/main/java/sg/dex/starfish/Asset.java [hunk 1]
```diff
@@ -90,4 +90,5 @@ public default DataAsset asDataAsset() {
 	 */
 	public Map<String, Object> getParamValue();
 
+    boolean isBundledAsset();
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/Asset.java]: I prefer the name "isBundle"

File: src/main/java/sg/dex/starfish/impl/AAsset.java [hunk 2]
```diff
@@ -71,5 +71,11 @@ public String getMetadataString() {
 	@Override
 	public abstract boolean isDataAsset();
 
+	@Override
+	public  boolean isBundledAsset(){
+		return getMetadata().get("type").toString().equalsIgnoreCase("bundle");
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/AAsset.java]: Use exact equals, we don't want to encourage or allow people to use different cases

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    public List<Asset> getAllSubAsset() {
+        return new ArrayList<>(subAssetMap.values());
+
+    }
+
+    public Asset getSubAssetById(String id) {
+        return subAssetMap.get(id);
+    }
+
+    public void addSubAsset(Asset asset) {
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: Shouldn't be a mutable class. If a new asset is added, need to create a new bundle instance

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    private static String buildMetaData(Map<String, Object> meta) {
+        // ToDO
+        String hash = Hex.toString(Hash.keccak256(new Random().toString()));
+
+        Map<String, Object> ob = new HashMap<>();
+        ob.put(DATE_CREATED, Instant.now().toString());
+        ob.put(CONTENT_HASH, hash);
+        ob.put(TYPE, "bundle");
+        ob.put(CONTENT_TYPE, "application/octet-stream");
+
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: Generated metadata should include the list of bundles assets as per DEP8

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+import java.time.Instant;
+import java.util.*;
+
+import static sg.dex.starfish.constant.Constant.*;
+
+public class MemoryAssetBundle extends AAsset {
+    private Map<String, Asset> subAssetMap;
+
+    private MemoryAssetBundle(String meta) {
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: Shouldn't need this constructor. Should never get created without a list of sub-assets

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import sg.dex.starfish.util.Hex;
+import sg.dex.starfish.util.JSON;
+
+import java.time.Instant;
+import java.util.*;
+
+import static sg.dex.starfish.constant.Constant.*;
+
+public class MemoryAssetBundle extends AAsset {
+    private Map<String, Asset> subAssetMap;
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: Should be `private final HashMap<String,Asset>()`. Also ensure a defensive copy is taken on creation (probably `new HashMap<>(source)` or something similar.

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        return false;
+    }
+
+    public List<Asset> getAllSubAsset() {
+        return new ArrayList<>(subAssetMap.values());
+
+    }
+
+    public Asset getSubAssetById(String id) {
+        return subAssetMap.get(id);
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: Probably need to inspect metadata here? Either:
- the subAssetMap must be guaranteed to contain all the assets listed in the metadata (class invariant upon construction). 
- or this method must inspect metadata to get sub assets, and add them to the map on demand

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    public boolean isDataAsset() {
+        return false;
+    }
+
+    @Override
+    public boolean isOperation() {
+        return false;
+    }
+
+    public List<Asset> getAllSubAsset() {
```
Comment (mikera) [inline src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: Should return a Map?

File: src/main/java/sg/dex/starfish/Asset.java (hunk 1 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/Asset.java]: doen

File: src/main/java/sg/dex/starfish/impl/AAsset.java (hunk 2 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/impl/AAsset.java]: done!

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java (hunk 3 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: done!

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java (hunk 5 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: done!

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java (hunk 7 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: need to discuss as this point is not very clear

File: src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java (hunk 8 above)
Comment (AyushVerma2) [reply src/main/java/sg/dex/starfish/impl/memory/MemoryAssetBundle.java]: done

**Pull request merged**